File: dart/poller.dart
// ─────────────────────────────────────────
// ALPR Poller — Dart
// Camera: Prama PT-NC140D3-WNM(D2)
// ─────────────────────────────────────────
// Install: https://dart.dev/get-dart
// Run:     dart run poller.dart

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

// ── Config (edit these)
const rtspUrl    = 'rtsp://{username}:{password}@{your_ip}:{port}/Streaming/channels/101';
const apiKey     = 'your key, if any';
const apiBase    = 'your api url';
const country    = 'IND';
const intervalMs = 2000;

const currFrame = '/tmp/alpr_curr.jpg';
const prevFrame = '/tmp/alpr_prev.jpg';

// ── Capture one JPEG frame via ffmpeg
Future<({Uint8List current, Uint8List? previous})> captureFrame() async {
  final curr = File(currFrame);
  final prev = File(prevFrame);

  if (await curr.exists()) {
    await curr.copy(prevFrame);
  }

  final result = await Process.run('ffmpeg', [
    '-loglevel', 'error',
    '-rtsp_transport', 'tcp',
    '-i', rtspUrl,
    '-frames:v', '1',
    '-q:v', '3',
    '-vf', 'scale=640:480',
    currFrame, '-y',
  ]);

  if (result.exitCode != 0) {
    throw Exception('ffmpeg exited ${result.exitCode}');
  }

  final current  = await curr.readAsBytes();
  final previous = await prev.exists() ? await prev.readAsBytes() : null;
  return (current: current, previous: previous);
}

// ── Motion detection: simple byte diff score
bool detectMotion(Uint8List current, Uint8List? previous) {
  if (previous == null) return false;

  final len = current.length < previous.length ? current.length : previous.length;
  var diff = 0.0;
  var samples = 0;
  for (var i = 0; i < len; i += 10) {
    diff += (current[i] - previous[i]).abs();
    samples++;
  }
  final score = (diff / samples) / 255.0 * 100.0;
  print('[motion] ${score.toStringAsFixed(2)}%');
  return score > 30;
}

// ── POST frame to ALPR API
Future<Map<String, dynamic>> scanPlate() async {
  final boundary = 'dartboundary${DateTime.now().millisecondsSinceEpoch}';
  final imageBytes = await File(currFrame).readAsBytes();

  final body = <int>[];
  void writeln(String s) => body.addAll(utf8.encode('$s\r\n'));

  writeln('--$boundary');
  writeln('Content-Disposition: form-data; name="image"; filename="frame.jpg"');
  writeln('Content-Type: image/jpeg');
  writeln('');
  body.addAll(imageBytes);
  writeln('');
  writeln('--$boundary');
  writeln('Content-Disposition: form-data; name="country"');
  writeln('');
  writeln(country);
  body.addAll(utf8.encode('--$boundary--\r\n'));

  final client = HttpClient();
  final uri    = Uri.parse('$apiBase/plate/recognize');
  final req    = await client.postUrl(uri);

  req.headers.set('x-api-key', apiKey);
  req.headers.set('Content-Type', 'multipart/form-data; boundary=$boundary');
  req.add(body);

  final res     = await req.close();
  final payload = await utf8.decodeStream(res);
  client.close();

  if (res.statusCode >= 300) throw Exception('ALPR API ${res.statusCode}');
  return jsonDecode(payload) as Map<String, dynamic>;
}

// ── Notify backend after confirmed plate
Future<void> notifyBackend(Map<String, dynamic> result) async {
  final client = HttpClient();
  final uri    = Uri.parse(apiBase.replaceAll('/api/v1', '/api/entry'));
  final req    = await client.postUrl(uri);

  req.headers.contentType = ContentType.json;
  req.headers.set('x-api-key', apiKey);
  req.write(jsonEncode({
    'plate':      result['plate'],
    'confidence': result['confidence'],
    'country':    country,
    'timestamp':  DateTime.now().toUtc().toIso8601String(),
    'source':     'dart-poller-v1',
  }));

  await (await req.close()).drain<void>();
  client.close();
}

// ── Main loop
Future<void> main() async {
  print('[alpr] dart poller started');
  print('[camera] $rtspUrl');

  var lastPlateTime     = DateTime.fromMillisecondsSinceEpoch(0);
  var consecutiveErrors = 0;
  final interval        = Duration(milliseconds: intervalMs);
  final cooldown        = const Duration(seconds: 8);

  while (true) {
    final t0 = DateTime.now();

    try {
      if (DateTime.now().difference(lastPlateTime) < cooldown) {
        print('[cooldown] skipping');
        await Future.delayed(interval);
        continue;
      }

      final (:current, :previous) = await captureFrame();
      consecutiveErrors = 0;

      if (!detectMotion(current, previous)) {
        print('[motion] none');
      } else {
        print('[motion] detected — scanning plate');
        final result = await scanPlate();
        print('[alpr] $result');

        final plate = result['plate'];
        if (plate != null && plate.toString().isNotEmpty) {
          print('[plate] $plate (${result['confidence']}%)');
          await notifyBackend(result);
          lastPlateTime = DateTime.now();
          print('[cooldown] 8s');
        } else {
          print('[plate] none found');
        }
      }
    } catch (e) {
      consecutiveErrors++;
      print('[error] #$consecutiveErrors: $e');
      if (consecutiveErrors >= 3) {
        print('[error] too many failures, waiting 30s');
        await Future.delayed(const Duration(seconds: 30));
        consecutiveErrors = 0;
      }
    }

    final elapsed = DateTime.now().difference(t0);
    if (elapsed < interval) await Future.delayed(interval - elapsed);
  }
}

File: bun/poller.ts
// ─────────────────────────────────────────
// ALPR Poller — Bun (TypeScript)
// Camera: Prama PT-NC140D3-WNM(D2)
// ─────────────────────────────────────────
// Install: curl -fsSL https://bun.sh/install | bash
// Run:     bun run poller.ts

// ── Config (edit these)
const RTSP_URL    = 'rtsp://{username}:{password}@{your_ip}:{port}/Streaming/channels/101';
const API_KEY     = 'your key, if any';
const API_BASE    = 'your api url';
const COUNTRY     = 'IND';
const INTERVAL_MS = 2000;

const CURR_FRAME = '/tmp/alpr_curr.jpg';
const PREV_FRAME = '/tmp/alpr_prev.jpg';

const sleep = (ms: number) => new Promise(r => setTimeout(r, ms));

// ── Capture one JPEG frame via ffmpeg
async function captureFrame(): Promise<{ current: Buffer; previous: Buffer | null }> {
  const curr = Bun.file(CURR_FRAME);
  if (await curr.exists()) {
    await Bun.write(PREV_FRAME, await curr.arrayBuffer());
  }

  const proc = Bun.spawn([
    'ffmpeg', '-loglevel', 'error',
    '-rtsp_transport', 'tcp',
    '-i', RTSP_URL,
    '-frames:v', '1', '-q:v', '3',
    '-vf', 'scale=640:480',
    CURR_FRAME, '-y',
  ], { stdout: 'ignore', stderr: 'ignore' });

  const code = await proc.exited;
  if (code !== 0) throw new Error(`ffmpeg exited ${code}`);

  const current  = Buffer.from(await Bun.file(CURR_FRAME).arrayBuffer());
  const prevFile = Bun.file(PREV_FRAME);
  const previous = (await prevFile.exists())
    ? Buffer.from(await prevFile.arrayBuffer())
    : null;

  return { current, previous };
}

// ── Motion detection: MD5 hash check + sampled byte diff
async function detectMotion(current: Buffer, previous: Buffer | null): Promise<boolean> {
  if (!previous) return false;

  const h1 = Bun.CryptoHasher.hash('md5', current,  'hex');
  const h2 = Bun.CryptoHasher.hash('md5', previous, 'hex');
  if (h1 === h2) return false;

  const len = Math.min(current.length, previous.length);
  let diff = 0;
  for (let i = 0; i < len; i += 10) diff += Math.abs(current[i] - previous[i]);
  const score = (diff / (len / 10)) / 255 * 100;
  console.log(`[motion] ${score.toFixed(2)}%`);
  return score > 30;
}

// ── POST frame to ALPR API
async function scanPlate(): Promise<Record<string, unknown>> {
  const form = new FormData();
  form.append('image', new Blob([await Bun.file(CURR_FRAME).arrayBuffer()], { type: 'image/jpeg' }), 'frame.jpg');
  form.append('country', COUNTRY);

  const res = await fetch(`${API_BASE}/plate/recognize`, {
    method: 'POST',
    headers: { 'x-api-key': API_KEY },
    body: form,
  });
  if (!res.ok) throw new Error(`ALPR API ${res.status}`);
  return res.json();
}

// ── Notify backend after confirmed plate
async function notifyBackend(result: Record<string, unknown>): Promise<void> {
  await fetch(`${API_BASE.replace('/api/v1', '')}/api/entry`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'x-api-key': API_KEY },
    body: JSON.stringify({
      plate:      result.plate,
      confidence: result.confidence,
      country:    COUNTRY,
      timestamp:  new Date().toISOString(),
      source:     'bun-poller-v1',
    }),
  });
}

// ── Main loop
console.log('[alpr] bun poller started');
console.log(`[camera] ${RTSP_URL}`);

let lastPlateTime     = 0;
let consecutiveErrors = 0;

while (true) {
  const t0 = Date.now();
  try {
    if (Date.now() - lastPlateTime < 8000) {
      console.log('[cooldown] skipping');
      await sleep(INTERVAL_MS);
      continue;
    }

    const { current, previous } = await captureFrame();
    consecutiveErrors = 0;

    if (!await detectMotion(current, previous)) {
      console.log('[motion] none');
    } else {
      console.log('[motion] detected — scanning plate');
      const result = await scanPlate();
      console.log('[alpr]', JSON.stringify(result));

      if (result.plate) {
        console.log(`[plate] ${result.plate} (${result.confidence}%)`);
        await notifyBackend(result);
        lastPlateTime = Date.now();
        console.log('[cooldown] 8s');
      } else {
        console.log('[plate] none found');
      }
    }
  } catch (err: unknown) {
    consecutiveErrors++;
    console.error(`[error] #${consecutiveErrors}: ${(err as Error).message}`);
    if (consecutiveErrors >= 3) {
      console.error('[error] too many failures, waiting 30s');
      await sleep(30_000);
      consecutiveErrors = 0;
    }
  }

  const elapsed = Date.now() - t0;
  if (elapsed < INTERVAL_MS) await sleep(INTERVAL_MS - elapsed);
}
